experimental/ir/lower_eval.go

```diff
@@ -0,0 +1,570 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ir
+
+import (
+	"fmt"
+	"math"
+	"math/big"
+	"strconv"
+	"strings"
+
+	"github.com/bufbuild/protocompile/experimental/ast"
+	"github.com/bufbuild/protocompile/experimental/ast/predeclared"
+	"github.com/bufbuild/protocompile/experimental/internal"
+	"github.com/bufbuild/protocompile/experimental/internal/taxa"
+	"github.com/bufbuild/protocompile/experimental/report"
+	"github.com/bufbuild/protocompile/experimental/seq"
+	"github.com/bufbuild/protocompile/experimental/token"
+	"github.com/bufbuild/protocompile/experimental/token/keyword"
+	"github.com/bufbuild/protocompile/internal/ext/mapsx"
+)
+
+const (
+	fieldNumberBits = 29
+	fieldNumberMax  = 1<<fieldNumberBits - 1
+	firstReserved   = 19000
+	lastReserved    = 19999
+)
+
+// evaluateFieldNumbers evaluates all non-extension field numbers: that is,
+// the numbers in reserved ranges and in non-extension field and enum value
+// declarations.
+func evaluateFieldNumbers(f File, r *report.Report) {
+	// TODO: Evaluate reserved ranges.
+
+	for ty := range seq.Values(f.AllTypes()) {
+		tags := make(map[int32]Field, ty.Fields().Len())
+		for field := range seq.Values(ty.Fields()) {
+			n, ok := evaluateFieldNumber(field, r)
+			field.raw.number = n
+			if !ok {
+				continue
+			}
+
+			// TODO: Need to check allow_alias here.
+			if first, ok := mapsx.Add(tags, n, field); !ok {
+				what := taxa.FieldNumber
+				if ty.IsEnum() {
+					what = taxa.EnumValue
+				}
+				r.Errorf("%ss must be unique", what).Apply(
+					report.Snippetf(field.AST().Value(), "used again here"),
+					report.Snippetf(first.AST().Value(), "first used here"),
+				)
+			}
+		}
+
+		// TODO: compare with extension/reserved ranges.
+	}
+}
+
+// evaluateExtensionNumbers evaluates all extension field numbers: that is,
+// the numbers on extension ranges and extension fields.
+func evaluateExtensionNumbers(f File, r *report.Report) {
+	// TODO: Evaluate extension ranges.
+
+	for extn := range seq.Values(f.AllExtensions()) {
+		n, _ := evaluateFieldNumber(extn, r)
+		extn.raw.number = n
+
+		// TODO: compare with extension ranges.
+	}
+}
+
+func evaluateFieldNumber(field Field, r *report.Report) (int32, bool) {
+	if field.AST().Value().IsZero() {
+		return 0, false // Diagnosed for us elsewhere.
+	}
+
+	e := &evaluator{
+		Context: field.Context(),
+		Report:  r,
+		scope:   field.FullName().Parent(),
+	}
+
+	// Don't bother allocating a whole Value for this.
+	v, ok := e.evalBits(evalArgs{
+		expr:   field.AST().Value(),
+		uint29: field.IsMessageField(), // TODO: MessageSet.
```

Comment:
I think this will call for a tri-state value, not just a bool flag.
1. int32 (enum values)
2. int31 (message-set extension numbers)
3. uint29 (field numbers and non-message-set extension numbers)